# Imports and Data Loading

In [1]:
import sys
import numpy as np
sys.path.append('../')
from src.models.baselines import PopularRecommender, SimpleJaccard, CosineKNN
import pandas as pd
from src.models.model import RedditGraph
from src.models.evaluator import Evaluator
import src.util.tigergraph_util as tgu
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv('../data/out/user_subreddit.csv')
df

product19           IAmA   1
0       RegularFreddieWilson           pics   2
1              superdarkness          funny  13
2                   tforcram           pics   1
3                      Misio          trees   5
4             greenstripe333            WTF   1
...                      ...            ...  ..
699817             shinratdr          Music   1
699818           geeforce272   IWantToLearn   1
699819              Maroonit       gonewild   1
699820           royaelliott         hockey   1
699821            LeeAmerica  entertainment   1

[699822 rows x 3 columns]

# Our user:
In this notebook, we will make some subreddit recommendations for user 'lowpass'.

In [5]:
user = 'lowpass'

# Evaluation:
Creating an evaluation helper:

In [2]:
evaluator = Evaluator('../data/out/test_interactions.csv', subset=50)

# Baseline recommender models:

## Popularity recommender:
Recommends popular subreddits user does not belong to

In [5]:
pop_rec = PopularRecommender(df)
pop_rec.recommend(user, n=5)

['politics', 'IAmA', 'trees', 'worldnews', 'science']

In [6]:
evaluator.precision_recall(pop_rec, 10)

(0.028386919483373657, 0.22462007703145379)

precision: 0.028386919483373657

recall:    0.22462007703145379

## Jaccard similarity recommender:
Recommends subreddits by determining what similar users (determined by Jaccard similarity) belong to

In [7]:
jaccard_rec = SimpleJaccard(df)
jaccard_rec.recommend(user)

['learnprogramming']

In [5]:
evaluator.precision_recall(jaccard_rec, 50)

NameError: name 'evaluator' is not defined

precision: 0.009563066776587017

recall:    0.07598568418848715

## Cosine similarity KNN recommender:
Recommends subreddits by determining what similar users (determined by nearest-neighbor similarity in a KNN) belong to

In [9]:
knn_rec = CosineKNN(df)
knn_rec.recommend(user)

['beer']

In [10]:
evaluator.precision_recall(knn_rec, 10)

(0.02346024330772864, 0.12531798374749745)

precision: 0.02346024330772864

recall:    0.12531798374749745

# Interaction graph recommender models:

In [3]:
reddit_graph = RedditGraph('../config/tigergraph.json')

## Louvain community recommender:
Recommends subreddits by determining what similar users (determined by assigned Louvain community) belong to

In [4]:
mdl_params = {'how': 'louvain'}
evaluator.precision_recall(reddit_graph, 10, **mdl_params)

ZeroDivisionError: division by zero

## Centrality KNN community recommender:
Recommends subreddits by determining what similar users (determined by nearest neighbors of the following centrality metrics: pagerank, louvain, label propagation, degree) belong to

In [5]:
reddit_graph.fit_knn('user', k=10)

169791 243808
169790


In [6]:
mdl_params = {'how': 'knn'}
evaluator.precision_recall(reddit_graph, 10, **mdl_params)

(0.01836734693877551, 0.1836734693877551)

In [2]:
conn = tgu.connection('../config/tigergraph.json')

In [3]:
loader = tgu.load_graph(conn, 64)

In [4]:
for batch in loader:
    print(batch)

Exception in thread Thread-4:
Traceback (most recent call last):
  File "c:\Users\Scott\AppData\Local\Programs\Python\Python39\lib\threading.py", line 954, in _bootstrap_inner
    self.run()
  File "c:\Users\Scott\AppData\Local\Programs\Python\Python39\lib\threading.py", line 892, in run
    self._target(*self._args, **self._kwargs)
  File "c:\Users\Scott\AppData\Local\Programs\Python\Python39\lib\site-packages\pyTigerGraph\gds\dataloaders.py", line 492, in _request_rest
    resp = tgraph.runInstalledQuery(
  File "c:\Users\Scott\AppData\Local\Programs\Python\Python39\lib\site-packages\pyTigerGraph\pyTigerGraphQuery.py", line 167, in runInstalledQuery
    return self._post(self.restppUrl + "/query/" + self.graphname + "/" + queryName,
  File "c:\Users\Scott\AppData\Local\Programs\Python\Python39\lib\site-packages\pyTigerGraph\pyTigerGraphBase.py", line 310, in _post
    return self._req("POST", url, authMode, headers, data, resKey, skipCheck, params)
  File "c:\Users\Scott\AppData\Loca